merlion/models/forecast/deep_models/deep_forecast_base.py

```diff
@@ -0,0 +1,340 @@
+            invert_transform=invert_transform,
+            **kwargs,
+        )
+        self.n_past = n_past
+
+        self.ts_freq = ts_freq
+        self.ts_encoding = ts_encoding
+        self.start_token_len = start_token_len
+
+
+class DeepForecaster(DeepModelBase, ForecasterBase):
+    """
+    Base class for a deep forecaster model
+    """
+
+    config_class = DeepForecasterConfig
+
+    def __init__(self, config: DeepForecasterConfig):
+        super().__init__(config)
+
+    def _create_model(self):
+
+        self.config.device = torch.device("cpu")
+
+        self.deep_model = self.deep_model_class(self.config)
+        print("I am at the creating model section")
+
+        # initialize optimizer
+        optim_dict = {
+            "adam": torch.optim.Adam,
+            "adamw": torch.optim.AdamW,
+            "sgd": torch.optim.SGD,
+            "adagrad": torch.optim.Adagrad,
+            "rmsprop": torch.optim.RMSprop,
+        }
+        if self.config.optim_name in optim_dict:
+            self.optimizer = optim_dict[self.config.optim_name](
+                self.deep_model.parameters(),
+                lr=self.config.lr,
+            )
+        else:
+            raise NotImplementedError("%s optimizer is not supported yet..." % (self.config.optim_name))
+
+        # initialize loss function
+        if self.config.criterion == "mse":
+            self.loss_fn = nn.MSELoss()
+        else:
+            raise NotImplementedError("%s loss is not supported..." % (self.config.criterion))
+
+        self._post_model_creation()
+
+    def _post_model_creation(self):
+        pass
+
+    def evaluate(self, evaluate_data: pd.DataFrame, metric: str = "mse"):
+        config = self.config
+        self.deep_model.eval()
+        eval_dataset = RollingWindowDataset(
+            evaluate_data,
+            n_past=config.n_past,
+            n_future=config.max_forecast_steps,
+            batch_size=config.batch_size,
+            target_seq_index=None,  # have to set None, we use config.target_seq_index later in the training
+            ts_encoding=config.ts_encoding,
+            ts_freq=config.ts_freq,
+            start_token_len=config.start_token_len,
+            flatten=False,
+            shuffle=False,
+        )
+
+        all_preds = []
+        all_trues = []
+
+        for i, batch in enumerate(eval_dataset):
+            with torch.no_grad():
+                loss, outputs, y_true = self._deep_batch_iter(batch, self.config)
+                pred = outputs.detach().cpu().numpy()
+                true = y_true.detach().cpu().numpy()
+                all_preds.append(pred)
+                all_trues.append(true)
+
+        preds = np.concatenate(all_preds, axis=0)
+        trues = np.concatenate(all_trues, axis=0)
+
+        logger.info("test shape:" + str(preds.shape) + str(trues.shape))
+
+        if metric == "mse":
+            pred_err = np.mean(np.sum((preds - trues) ** 2, axis=-1), axis=-1)
```

Comment:
Rename `metric` to `loss_fn`, and make it be an instance of a `LossFunction` enum as suggested above. Then, you can call `loss_fn.value(pred, true)` (see the implementation of `ForecastMetric` or `TSADMetric` for some inspiration). Note that you'll probably want to operate on torch tensors instead of numpy arrays here (though still on CPU), so you can use the call to compute the training loss.

Also, please rename `pred_err` to `loss`. This is because `pred_err` refers to the standard error of the prediction in many parts of the repo.